src/main.cc

```diff
@@ -444,6 +444,8 @@ load_cfg(void)
 		}
 	}
 
+	box_set_lua_memory();
```

Comment:
`src/lua` subsystem doesn't use `src/box/` at all. It looks like breaking architectural dependency directions.
I'll bring it to `box_lua_init()`. Is it ok?

Reply:
Raw thought: a `box.cfg` option was not requested. And the given option has no relation to the database at all. We can expose a Lua function to set the limit from some module in `src/lua`.

`tarantool` module is informational, not a good option. Create some `tarantool.vm` module? Don't know.

Any ideas?

Reply:
There is an idea to place the Lua API to the `misc` module.

```yaml
# tarantool> misc
---
- getmetrics: 'function: builtin#203'
  sysprof:
    stop: 'function: builtin#205'
    report: 'function: builtin#206'
    start: 'function: builtin#204'
  memprof:
    start: 'function: builtin#207'
    stop: 'function: builtin#208'
...
```

It looks as a good place. However, the implementation of this module is in the luajit repository (our fork). If we're going to expose a function from it, it seems that the code should reside in the luajit fork too[^1].

I like the idea to have this machinery in the luajit codebase, because luajit's memory limit seems closer to luajit itself rather than tarantool. @Buristan, what do you think?

[^1]: Technically we can add a function into the module table from tarantool, but I don't like that the implementatoin of the module would spread across different files and even repositories in this case.

Reply:
We discussed this with @Totktonada offline and came to the following conclusion:

Decalring this custom allocator function inside the LuaJIT fork is hardly maintainable (changes in the Tarantool allocator require changes in the LuaJIT fork that look inconsistent) and too specific to be a part of the API, so it should be done as an __internal__ module inside Tarantool.

So we should provide an internal module (say <src/lua/alloc.c>) that provides two Lua C functions -- to set the alloc function and restore the previous one. Also, it provides in the Lua namespace function `internal.alloc.setlimit()` which allows to easily set the limit of the Lua memory.

Reply:
Applied the suggested approach.

Reply:
In that case, why do we need to call `box_set_lua_memory()` here (I thought that it should be called in some other place, shouldn't it)?

Reply:
What I mean is that it is just an unexpected place for me (I suppose we should end managing with the `pid_file` first). OTOH, it should be done before `box_init_say()`, since it uses Lua memory.
I will ask @Totktonada for the final judgment here.

Reply:
I remember a result of our past discussions as follows:

* We create an internal Lua module (`internal.alloc`) with the allocator and functions to enable/disable it and get/set the memory limit.
* Tarantool enables the allocator for the main Lua state early at startup (and the default 2 GiB limit is set).
* We create a configuration option and a configuration applier (`src/box/lua/config/applier/lua.lua` or kinda) that allows to configure the limit (it uses the `internal.alloc` API directly).

This way we don't need to add any `box.cfg` option and it looks meaningful -- this option is not related to box (the database). We have fiber slice limit configuration implemented this way.

BTW, I propose to handle decreasing of the limit in the following way. If the limit is decreased, a warning is issued and the limit remains unchanged (until restart). It is how non-dynamic `box.cfg` options are handled.

https://github.com/tarantool/tarantool/blob/d33b14793bea58520139172599621daaaf66b947/src/box/lua/config/applier/box_cfg.lua#L550-L558

The motivation here is that sometimes it is convenient to setup needed options first and restart instances then.

Reply:
Thank you for the suggestion! Now I completely agree with you the option is not related to box at all. I am moving it to a new config applier.

Reply:
Seems solved (raised a separate discussiong about the non-dynamism).